index.go

```diff
@@ -0,0 +1,171 @@
+package opensearchtools
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+)
+
+// Index defines a method which knows how to make an OpenSearch [Index] request.
+// It should be implemented by a version-specific executor.
+//
+// [Index]: https://opensearch.org/docs/latest/api-reference/index-apis/index/
+type Index interface {
+	Index(ctx context.Context, req *IndexRequest) (OpenSearchResponse[IndexResponse], error)
+}
+
+type IndexRequest struct {
+	Operation IndexOperation
+	Routing   string
+}
+
+type IndexOp string
+
+const (
+	// IndexCreate creates an index and returns an error otherwise.
+	IndexCreate IndexOp = "create"
+
+	// IndexDelete delete an index if it exists and returns an error otherwise.
+	IndexDelete IndexOp = "delete"
+
+	// IndexGet gets an index if it exists and returns an error otherwise.
+	IndexGet IndexOp = "get"
+
+	// IndexExists checks the existence of an index
+	IndexExists IndexOp = "exists"
+)
+
+// IndexOperation various different index operations to be executed
+type IndexOperation struct {
+	Operation IndexOp
+	Indices   []string
+	Config    *indexRequestConfig
+	Doc       RoutableDoc
+}
+
+func NewIndexCreateOperation(index string, doc RoutableDoc, ops ...IndexReqOptionFunc) *IndexOperation {
+	config := &indexRequestConfig{}
+	for _, o := range ops {
+		o(config)
+	}
+	return &IndexOperation{Indices: []string{index}, Doc: doc, Config: config, Operation: IndexCreate}
+}
+
+func NewIndexDeleteOperation(index []string, ops ...IndexReqOptionFunc) *IndexOperation {
+	config := &indexRequestConfig{}
+	for _, o := range ops {
+		o(config)
+	}
+	return &IndexOperation{Indices: index, Config: config, Operation: IndexCreate}
+}
+
+func NewIndexGetOperation(index []string, ops ...IndexReqOptionFunc) *IndexOperation {
+	config := &indexRequestConfig{}
+	for _, o := range ops {
+		o(config)
+	}
+	return &IndexOperation{Indices: index, Config: config, Operation: IndexGet}
+}
+
+func NewIndexExistOperation(index []string, ops ...IndexReqOptionFunc) *IndexOperation {
+	config := &indexRequestConfig{}
+	for _, o := range ops {
+		o(config)
+	}
+	return &IndexOperation{Indices: index, Operation: IndexExists}
+}
+func (i *IndexOperation) OperationType() IndexOp {
+	return i.Operation
+}
+
+func (i *IndexOperation) GetIndices() []string {
+	return i.Indices
+}
+
+func (i *IndexOperation) GetDoc() RoutableDoc {
+	return i.Doc
+}
+
+type indexRequestConfig struct {
+	masterTimeout         time.Duration
+	clusterManagerTimeout time.Duration
+	expandWildcards       string // make this enum later
+	waitForActiveShards   string
+	allowNoIndices        bool
+}
+
+// IndexReqOptionFunc for functional options
+type IndexReqOptionFunc func(config *indexRequestConfig)
+
+// WithMasterTimeOut to add the master timeout
+func WithMasterTimeOut(d time.Duration) IndexReqOptionFunc {
```

Comment:
This might be the biggest point. I'm not opposed to the config pattern. What I don't love is that this creates inconsistence in our request objects. All of our other requests uses a builder pattern. I would argue that this one should also follow the builder pattern. Or we should update all the other requests to use this config pattern.

@dylan-bourque , @jeffmclean-cs , thoughts?

Reply:
True, that's the point I wanted to discuss as well.

Reply:
I'd prefer to stick with the Builder pattern we have with the other request types. Makes sense to me to be consistent. Personally I prefer the builder pattern over this (@cs-marshall called it the config pattern, but I think it's more commonly called the functional options pattern in Go world), however that is just personally preference and it looks like this functional options pattern is more common in Go so I'd be willing to negotiate if other people have strong reasons for it. Keep in mind it would be a breaking change at this point and involve changes to consumer code.

Reply:
I actually like functional options and use them often but I think that consistency across the library is more important.

Reply:
I also like the functional options, reminds me the old Kotlin and Rust days :) But yeah I'll update it with the builder pattern.